### PR TITLE
aws-gate: init at 0.11.3

### PR DESCRIPTION
**Pull request by tirimia**
## Description of changes

Added `aws-gate`, a session manager that makes working with aws a lot smoother.
In the process of adding it, I also had to add the `unix-ar` python package as it was a runtime dependency not found in nixpkgs.
Left a comment regarding the necessity of `unix-ar` in the `aws-gate` recipe. Still relatively new to nix so not sure if my suggestion makes complete sense, looking forward to feedback 😄

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by KsiXy — APPROVED** [review]

File: pkgs/tools/admin/aws-gate/default.nix
Comment (aszlig) [inline pkgs/tools/admin/aws-gate/default.nix:33]: That's probably a good idea if you don't want to maintain `unix-ar`. When looking at the `aws-gate` source, it seems that the whole `aws_gate.bootstrap` module only is used in tests and for invoking the `bootstrap` function here:

https://github.com/xen0l/aws-gate/blob/52f7281af2805e28d44ee3906b14d48acb5201fb/aws_gate/cli.py#L271

So in theory, it should be possible to just remove the whole `bootstrap` subcommand from the `ArgumentParser` and remove all related modules (and tests) and it should work fine.

File: pkgs/top-level/all-packages.nix
Comment (aszlig) [inline pkgs/top-level/all-packages.nix:3402]: Btw. we are going to replace `all-packages.nix` by a new `by-name` structure as detailed in <https://github.com/NixOS/rfcs/blob/c65c8321782b40844167beb48818471f70900d9d/rfcs/0140-simple-package-paths.md>, so you don't need to use `callPackage` here and simply move the `aws-gate` over to `pkgs/by-name/aw/aws-gate/package.nix`

File: pkgs/tools/admin/aws-gate/default.nix
Comment (tirimia) [reply pkgs/tools/admin/aws-gate/default.nix:33]: I am not opposed to maintaining `unix-ar`, doubt there will be more versions to maintain in the future and I have found quite a few repositories on github referencing it. When those eventually make their way to nixpkgs, having the dependency here already would ease their maintainers' work.

Still believe the dependency should be removed from `aws-gate` though; do I make a patch to remove the command completely or should I keep the command and have it simply print that the package installed from nixpkgs will always have the plugin? 
I'm leaning towards the latter since it would not perturb user-facing functionality. However, there I am not sure how things are usually done when porting such packages to nix so I defer to your judgment.

File: pkgs/top-level/all-packages.nix
Comment (tirimia) [reply pkgs/top-level/all-packages.nix:3402]: Oh, that's great to hear, will do it asap. Thank you for taking the time to review the PR 😸

File: pkgs/by-name/aw/aws-gate/disable-bootstrap.patch
Comment (aszlig) [inline pkgs/by-name/aw/aws-gate/disable-bootstrap.patch:5]: I'd probably not include the file deletions in the patch but rather as `postPatch` in `mkDerivation` like suggested below, because every minor upstream change in these fails will cause the patch to not apply anymore.

File: pkgs/by-name/aw/aws-gate/package.nix
Comment (aszlig) [inline pkgs/by-name/aw/aws-gate/package.nix:22]: As mentioned in the above comment, the `postPatch` hook could look like this:

```suggestion

  postPatch = ''
    rm aws_gate/bootstrap.py tests/unit/test_bootstrap.py
  '';

```

File: pkgs/tools/admin/aws-gate/default.nix
Comment (tirimia) [reply pkgs/tools/admin/aws-gate/default.nix]: Got that working and in the process also found that the `requests` dependency could also be removed if no bootstrapping is needed 🥳

File: pkgs/by-name/aw/aws-gate/disable-bootstrap.patch
Comment (tirimia) [reply pkgs/by-name/aw/aws-gate/disable-bootstrap.patch]: That would indeed have been very unpleasant. Fixed!

File: pkgs/by-name/aw/aws-gate/package.nix
Comment (tirimia) [reply pkgs/by-name/aw/aws-gate/package.nix:23]: Implemented exactly like that

**Review by aszlig — APPROVED** [review]

**Comment by aszlig** [comment]
Thanks and congrats on your first contribution :smiley:

**Comment by tirimia** [comment]
Thank you for all the support ❤️ It's an honor to contribute to such an awesome project, I'll be back with more PRs!

**Pull request merged**
